Add `TracingStore` + `TracingVersionStore` facades

snazy opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/projectnessie/nessie/986)
<!-- Reviewable:end -->

laurentgo reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/986)* status: 0 of 17 files reviewed, 2 unresolved discussions (waiting on @snazy)

___
*[versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java, line 64 at r1](https://reviewable.io/reviews/projectnessie/nessie/986#-MWfGBJaBn-rK103OCNz:-MWfGBJaBn-rK103OCO-:b-axp298) ([raw file](https://github.com/projectnessie/nessie/blob/6aaa3c3c885fb57a75ddc156b8eab5c821c7d66f/versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java#L64)):*
> ```Java
>    */
>   @VisibleForTesting
>   public TracingVersionStore(VersionStore<VALUE, METADATA> delegate, Supplier<Tracer> tracerSupplier) {
> ```

does it need to be public or would package protected visibility be enough?
___
*[versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java, line 326 at r1](https://reviewable.io/reviews/projectnessie/nessie/986#-MWfHnn30504t89NyW3h:-MWfHnn30504t89NyW3i:bmy87i8) ([raw file](https://github.com/projectnessie/nessie/blob/6aaa3c3c885fb57a75ddc156b8eab5c821c7d66f/versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java#L326)):*
> ```Java
> 
>   private static RuntimeException traceRuntimeException(Scope scope, RuntimeException e) {
>     Tags.ERROR.set(scope.span().log(ImmutableMap.of(Fields.EVENT, Tags.ERROR.getKey(),
> ```

is it a good pattern or is it similar to how we should not log an error and throw back the exception which might not have been an error after all if the caller knows how to handle it?


<!-- Sent from Reviewable.io -->

snazy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/projectnessie/nessie/986)* status: 0 of 18 files reviewed, 2 unresolved discussions (waiting on @laurentgo)

___
*[versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java, line 64 at r1](https://reviewable.io/reviews/projectnessie/nessie/986#-MWfGBJaBn-rK103OCNz:-MWi2Oxw0JcnQHHSct4a:b-177il) ([raw file](https://github.com/projectnessie/nessie/blob/6aaa3c3c885fb57a75ddc156b8eab5c821c7d66f/versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java#L64)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

does it need to be public or would package protected visibility be enough?
</blockquote></details>

not really, no. changed it.
___
*[versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java, line 326 at r1](https://reviewable.io/reviews/projectnessie/nessie/986#-MWfHnn30504t89NyW3h:-MWi2TLjBAARmqAk00Sd:b-k09cgz) ([raw file](https://github.com/projectnessie/nessie/blob/6aaa3c3c885fb57a75ddc156b8eab5c821c7d66f/versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java#L326)):*
<details><summary><i>Previously, laurentgo (Laurent Goujon) wrote…</i></summary><blockquote>

is it a good pattern or is it similar to how we should not log an error and throw back the exception which might not have been an error after all if the caller knows how to handle it?
</blockquote></details>

It's a bit tricky IMO. On the one hand we have unchecked exceptions that indicate a real technical failure but also usages of e.g. `IllegalArgumentException` that indicate a user-error, but are returned as a "HTTP/500 server error" to clients. This particular one is meant to mark the trace as "error" because of some technical error. TL;DR it's hard to distinguish a real technical error from a user-error here. It feels like it's worth to have a discussion how to properly differentiate between "real technical runtime-exceptions" and actual "parameter validation exceptions"?


<!-- Sent from Reviewable.io -->

snazy commented:
Rebased + force-pushed due to conflicts

rymurr commented on versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java:
```diff
@@ -0,0 +1,334 @@
+  @Nonnull
+  public Hash toHash(@Nonnull NamedRef ref) throws ReferenceNotFoundException {
+    return delegate1Ex("VersionStore.toHash", b -> b
+        .withTag("nessie.version-store.operation", "ToHash")
```
I thin the tag names should be `static final String` s

snazy replied:
Done

rymurr commented on versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java:
```diff
@@ -0,0 +1,334 @@
+  }
+
+  @FunctionalInterface
+  interface Delegate<R> {
```
The overloading of `Delegate`, `delegate()` and `delegate` is slightly confusing. Could we rename a few to create more distinction?

snazy replied:
Renamed some of those.

rymurr commented on versioned/tiered/dynamodb/src/main/java/org/projectnessie/versioned/dynamodb/DynamoStore.java:
```diff
@@ -570,4 +598,14 @@ private static void verifyKeySchema(TableDescription description) {
         + "attribute with the name 'id'.", description.tableName()));
   }
+
+  private Tracer getTracer() {
```
we can only use the global tracer here but above we have the option to specify a tracer. Can we do one or the other? I prefer just using the global tracer.

snazy replied:
It's been due to `GlobalTracer` not allowíng us to register the "global tracer" once, but never change it. I've updated the tests to use a proxy-instance between the `Tracer` for the tests and `GlobalTracer`

rymurr commented on versioned/tiered/dynamodb/src/main/java/org/projectnessie/versioned/dynamodb/metrics/TracingExecutionInterceptor.java:
```diff
@@ -76,11 +77,11 @@ public void afterMarshalling(final AfterMarshalling context,
     final Span span = executionAttributes.getAttribute(SPAN_ATTRIBUTE);
     final SdkHttpRequest httpRequest = context.httpRequest();
 
```
nit: This class only exists while we are watiing for opentracing to update in Quarkus. Wonder if its worth maintaining cosmetic changes?

snazy replied:
ah, interesting. i've reverted the changes to this file.

rymurr commented on versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/InternalBranch.java:
```diff
@@ -403,57 +410,71 @@ private void save(Store store) {
+        try {
+          UpdateState updateState = initialState;
+          for (int attempt = 0; attempt < attempts; attempt++) {
+            try (Scope innerScope = createSpan("Attempt-" + attempt).startActive(true)) {
```
again minor concern over cost of this op vs payoff

snazy replied:
Yea, that particular one is tricky. The code is "too far away" from knowing whether the user wants things to be traced, but if you trace, the information is pretty useful.

rymurr commented on versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/store/TracingStore.java:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+
+  private static <C extends BaseValue<C>> String safeName(ValueType<C> type) {
```
some dup code here and in version store. Possible to create a util?

snazy replied:
Yup, done. Also tried to "unify" the test-code stuff to reuse common things and to make the test-code a more readable.

rymurr commented on versioned/tiered/tiered-impl/src/main/java/org/projectnessie/versioned/impl/InternalBranch.java:
```diff
@@ -403,57 +410,71 @@ private void save(Store store) {
-            LOGGER.debug("Completed collapse update on attempt {}.", attempt);
-            return producer.build().getBranch();
+      try (Scope outerScope = createSpan("InternalBranch.collapseIntentionLog")
+          .withTag("nessie.operation", "CollapseIntentionLog")
```
here and elsewhere: tag names could move to static final Strings

snazy replied:
done

rymurr reviewed (commented):
LGTM. The only comment is if we can try and abstract the tracing in `InternalBranch` and `DynamoStore`. I feel like its a bit harder to read and I don't fully understand the perf  implications

snazy commented:
Added a commit to address this. Opted to introduce an _internal_ `TieredVersionStoreConfig` object that carries the wait-for-collapse, enable-tracing and the two commit-attempts values, as adding another param to TVS's ctor felt too much.
Refactored the InternalBranch.UpdateState.collapse... method a bit to make it more readable.
